### PR TITLE
[AArch64][SelectionDAG] Lower multiplication by a constant to shl+sub+shl+sub

### LINKED ISSUE #89430: [AArch64] Suboptimal code for multiplication by certain constants
For some constants, GCC is able to generate sequences of `add` where LLVM generates `mul`. I have checked all constants between 1 and 100 (https://godbolt.org/z/rxej44fGj):

For all of the examples below(11, 13, 19, 21, 25, 27, 35, 37, 41, 49, 51, 69, 73, 81, 85), LLVM generates

```asm
mulK:
        mov     w8, K
        mul     w0, w0, w8
        ret
```

```asm
mul11:
        add     w1, w0, w0, lsl 2
        add     w0, w0, w1, lsl 1
        ret
```

```asm
mul13:
        add     w1, w0, w0, lsl 1
        add     w0, w0, w1, lsl 2
        ret
```

```asm
mul19:
        add     w1, w0, w0, lsl 3
        add     w0, w0, w1, lsl 1
        ret
```

```asm
mul21:
        add     w1, w0, w0, lsl 2
        add     w0, w0, w1, lsl 2
        ret
```

```asm
mul25:
        add     w0, w0, w0, lsl 2
        add     w0, w0, w0, lsl 2
        ret
```

```asm
mul27:
        add     w0, w0, w0, lsl 1
        add     w0, w0, w0, lsl 3
        ret
```

```asm
mul35:
        add     w1, w0, w0, lsl 4
        add     w0, w0, w1, lsl 1
        ret
```

```asm
mul37:
        add     w1, w0, w0, lsl 3
        add     w0, w0, w1, lsl 2
        ret
```

```asm
mul41:
        add     w1, w0, w0, lsl 2
        add     w0, w0, w1, lsl 3
        ret

```

```asm
mul49:
        add     w1, w0, w0, lsl 1
        add     w0, w0, w1, lsl 4
        ret
```

```asm
mul51:
        add     w0, w0, w0, lsl 1
        add     w0, w0, w0, lsl 4
        ret
```

```asm
mul69:
        add     w1, w0, w0, lsl 4
        add     w0, w0, w1, lsl 2
        ret
```

```asm
mul73:
        add     w1, w0, w0, lsl 3
        add     w0, w0, w1, lsl 3
        ret
```

```asm
mul81:
        add     w0, w0, w0, lsl 3
        add     w0, w0, w0, lsl 3
        ret
```

```asm
mul85:
        add     w0, w0, w0, lsl 2
        add     w0, w0, w0, lsl 4
        ret
```

### PR DESCRIPTION
**Pull request by vfdff**
Change the costmodel to lower a = b * C where C = 1 - (1 - 2^m) * 2^n to
              sub  w8, w0, w0, lsl #m
              sub  w0, w0, w8, lsl #n
Fix https://github.com/llvm/llvm-project/issues/89430

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/mul_pow2.ll [hunk 1]
```diff
@@ -615,6 +632,42 @@ define i32 @test97_fast_shift(i32 %x) "target-features"="+alu-lsl-fast" {
...
+
+  %mul = mul nsw i32 %x, 125 ; 63 = 1 - ((1-32) << 2)
+  ret i32 %mul
+}
+
+; Negative test: The shift number 5 is out of bound
+define i32 @test225_fast_shift(i32 %x) "target-features"="+alu-lsl-fast" {
+; CHECK-LABEL: test225_fast_shift:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov w8, #225 // =0xe1
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/AArch64/mul_pow2.ll:657]: You can write `x*225` as `x*-15*-15`.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -17611,6 +17628,8 @@ static SDValue performMulCombine(SDNode *N, SelectionDAG &DAG,
     //     => MV = (add (shl x, M), x); (add (shl MV, N), MV)
     // (mul x, (2^M + 1) * 2^N + 1))
     //     =>  MV = add (shl x, M), x); add (shl MV, N), x)
+    // (mul x, 1 - (1 - 2^M) * 2^N))
+    //     =>  MV = sub (x - (shl x, M)); add (x - (shl x, M))
```
Comment (efriedma-quic) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17632]: Formula looks inconsistent with other formulas.

File: llvm/test/CodeGen/AArch64/mul_pow2.ll (hunk 1 above)
Comment (vfdff) [reply llvm/test/CodeGen/AArch64/mul_pow2.ll:657]: oh, Yes.  225 = (1-16) * (1-16), which is not supported  now.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 2 above)
Comment (vfdff) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17632]: Updated, it is `MV = sub (x - (shl x, M)); add (x - (shl MV, N))`, thanks.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**